### ISSUE TITLE
[Permissions] [Idea] Support Android 12 handling of coarse/fine location permission

### ISSUE THREAD

**Issue by alexvanyo** [google/accompanist#819]
If an app is [targeting Android 12 or higher](https://developer.android.com/training/location/permissions#approximate-request), `ACCESS_FINE_LOCATION` can only be requested in combination with `ACCESS_COARSE_LOCATION`, and the user might only grant the application access to coarse location. If the user only granted approximate location, it is also possible to [ask to upgrade to precise location later](https://developer.android.com/training/location/permissions#upgrade-to-precise).

The current behavior from Accompanist (using `MultiplePermissionsState` while passing both types of location access), doesn't quite handle this scenario correctly, since the permission granted content will only be shown if both types of location access are granted. If the user has only granted approximate location, the "not granted" content or "not available" content will be shown instead.

**Comment by svenjacobs** [comment]
I'm also facing this issue and described my problem [here](https://github.com/google/accompanist/discussions/861).

**Comment by OliverRhyme** [comment]
Badly needed this to be on compose, or at least a reasonable workaround

**Comment by manuelvicnt** [comment]
Hi there! 👋 After giving this a go, I'm wondering which APIs you're missing. 

I believe the problem keeps lying on not being able to differentiate whether the permissions have been requested before or the user completely denied them (related to #549). IIRC, saving whether or not you've requested a permission before in disk is not a good solution since permissions can be reset automatically by the system or changed manually by the user.

I created this [gist with an implementation of requesting coarse/fine location permissions using the current APIs](https://gist.github.com/manuelvicnt/f34f769b23b5b68ab81274db95cb7d15). As you can see, I'm not differentiating the scenario presented above, I'm always leaving the Button enabled even though it doesn't do anything when the the user doesn't want to be asked again (i.e. permission completely denied). I hope we can get better UX guidance on this.

In a nutshell, you can quickly calculate if all permissions have been revoked in order to know if `COARSE` is granted but `FINE` is not.

PTAL and let me know what you think.

**Comment by svenjacobs** [comment]
Hello @manuelvicnt, please also see my [post](https://github.com/google/accompanist/discussions/861) where I describe a use case that is currently not possible.

**Comment by bubenheimer** [comment]
@manuelvicnt I gave this a go as well. The inability to differentiate between never-asked-before & completely-denied is indeed a big part of the problem here. However, this is not a new problem, app developers have had to work around it for years. The Android framework permissions APIs, most unfortunately, have made very opinionated API design decisions that cause poor UX and lead to complex code logic when a permission is *required* for an app to function; the situation is exacerbated when an app requires precise location for core functionality, e.g. for navigation or to locate the user/phone.

What *is* new in the case of accompanist-permissions is an absence of callbacks & insufficient state change information. This essentially prevents showing the right information to the user at the right time, because there is no signal when the user has dismissed the permissions dialog or when the system has resolved the permissions request on its own one way or another. The permissions state just does not really change that much, so no robust action can be taken in certain cases.

In the case of a permission that is required for core app functionality there is no way around offering the user access to app settings in case permission requests are exhausted, because what's the alternative? Hang/close/crash the app or tell the user to go away and reinstall?

The only halfway reasonable & robust way I have found for the required-permission scenario with accompanist-permissions: offer the user all available options whenever insufficient permissions are detected, and let the user make the right choice: show system permissions request dialog, go to settings, or cancel & exit the app. This sorta works, but it is rather poor UX, and falls short of the guided workflow that I can provide with a callback-based approach. For example, when the user picks "Ask every time" in the system permissions dialog, they now get this extra dialog every time.

**Comment by bubenheimer** [comment]
I tried this once more after noticing the recent addition of the 'onPermissionsResult` callback lambda to `rememberMultiplePermissionState()` in the 0.24.3 alpha. This change enables permission handling for the coarse/fine use case at par with procedural approaches, AFAICT. The trick is this:

    val locationPermissions: List<String> = listOf(ACCESS_COARSE_LOCATION, ACCESS_FINE_LOCATION)

    @Stable
    interface LocationPermissionsState : MultiplePermissionsState {
        /**
         * Supplies a well-defined measure of time/progression
         */
        val requestCount: UInt
    }

    @Composable
    fun rememberLocationPermissionsState(): LocationPermissionsState {
        val requestCountState: MutableState<UInt> = remember { mutableStateOf(0u) }

        val multiplePermissionsState: MultiplePermissionsState =
            rememberMultiplePermissionsState(locationPermissions) {
                if (it.isEmpty()) {
                    // BUG in accompanist-permissions library upon configuration change
                    return@rememberMultiplePermissionsState
                }

                requestCountState.value++
            }

        return object : LocationPermissionsState, MultiplePermissionsState by multiplePermissionsState {
            override val requestCount: UInt by requestCountState
        }
    }

`requestCount` makes it possible to react to an absence of change in permissions state after a permissions request. This is critical because once location permission requests are exhausted there is no change in permissions state after requesting permissions: the code has no idea whether the permissions request has completed, it's essentially in limbo. 

Depending on implementation details of the system permissions API this absence of permissions state change may also occur in other situations, for example, it may happen on older API levels where retries occur without a change in permissions state.

`requestCount` (or something equivalent) should be added to the accompanist-permissions API, as it is essential for keeping track of temporal progression.

N.B. the new callback mechanism has a bug on the library side as noted above where the callback is intermittently invoked with an empty permissions map upon configuration change. Possibly a race between the previous MultiplePermissionsState before the configuration change and the new one after, callbacks from the former not completely done yet? Just a guess.

@svenjacobs pretty sure this will enable you to address your use case.

**Comment by svenjacobs** [comment]
Thanks @bubenheimer for your detailed investigation. As time progressed, I'm currently working on other parts of my app. But I will have a look at this in the future.
